Add whiteboard for ClassFileTransformer services

tjwatson opened the pull request.

tjwatson committed:
Add whiteboard for ClassFileTransformer services

This allows a system transformer to be registered that applies
to all AppClassLoaders that get created for the server.

To be recognized as a system transformer the ClassFileTransformer
must be registered with the service property:

io.openliberty.classloading.system.transformer=true

System transformers differ from other transformers
added to the AppClassLoader class with
AppClassLoader.addTransformer(ClassFileTransformer). System
transformers can transform classes even if they were previously
stored in the shared classes cache.  But if a system transformer
does change a class bytes then the transformed class is not
stored in the shared classes cached, just like other transformers

tjwatson commented:
#build

tjwatson commented:
#build

dazavala commented on dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java:
```diff
@@ -319,42 +329,60 @@ boolean removeTransformer(ClassFileTransformer transformer) {
+                }
+            } else {
+                if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                    Tr.debug(tc, "Attempt to transform " + name + " from shared class cache because we have global tranformers");
```
Is the term "global" correct here?  The debug msg may be confusing because it does not indicate transformation is attempted using _only_ the system transformers.  I suspect we'll reuse this pattern for transformers added programmatically.

tjwatson replied:
No, I had started with a variable named `global` but then decided to rename it to `system` because that seemed a better name.  I'll update the debug message.

dazavala reviewed (commented):
@tjwatson I didn't find any issues regarding the re-factorization of the transform logic and class caching, including the new transformation of cached classes.  

The byte transformation debug msgs changed slightly; I added a comment regarding serviceability.  

Scanning the code base, I believe you passed a system transformer list to all necessary app classsloader types through all paths to their instantiation.  That alleviated my initial concern for NPEs in AppClassLoader.  The UnifiedClassLoader has a use-case that involves Spring class weaving.  We might revisit UCL in this regard, only.

I like the use of CopyOnWriteArrayList, though I can't opine on its overhead as I don't know how many services/tranformers will register.  

There are a few copyrights that need update; Libby will whine about that.  

Lastly, my understanding is that a beta fence should throw an exception when running in non-beta images and emit a single info message in beta images.  This solution enables new behavior for beta images, only, which I think is sufficient as the beta fence I described should be added to the feature/service which uses the whiteboard.

Approved since I found no required changes other than copyrights, which libby will handle.

tjwatson commented:
> @tjwatson I didn't find any issues regarding the re-factorization of the transform logic and class caching, including the new transformation of cached classes.
> 
> The byte transformation debug msgs changed slightly; I added a comment regarding serviceability.

I will fix this up.

> 
> Scanning the code base, I believe you passed a system transformer list to all necessary app classsloader types through all paths to their instantiation. That alleviated my initial concern for NPEs in AppClassLoader. The UnifiedClassLoader has a use-case that involves Spring class weaving. We might revisit UCL in this regard, only.

Good observation.  I missed that the ThreadContextClassLoader -> UnifiedClassLoader hierarchy is different than the AppClassLoader -> ContainerClassLoader hierachy.  Because UnifiedClassLoader is purely a delegating class loader (as far as I can tell).  It does not store any of the transformers itself.  Instead it delegates to other loaders (like the AppClassLoader) that will contain the transformers.  I don't think we need to do anything special hear for the system transformers.

With that said.  I do not know how any of our tranformers work for the AppClassLoader.getThrowawayClassLoader.  This is an instance of ShadowClassLoader.  When this class loader is used to define classes I see nowhere that it will call any of our transformers.  That includes the system transformers and the ones added with the addTransformer method.  Perhaps this is by design, but it is not clear at all in the code.  The Javadoc of the ShadowClassLoader does not really help clarify the concern either.  I expect it is fine to not transform because I think the ShadowClassLoader is only for introspecting the classes in such a way that the classes can be discarded.  Perhaps it is not important for the classes to be transformed for introspecting.

> 
> I like the use of CopyOnWriteArrayList, though I can't opine on its overhead as I don't know how many services/tranformers will register.

CopyOnWriteArrayList is great for when you have many reads vs writes to the list.  The read operation is nearly identical to a normal ArrayList.  The writes are much more expensive because of concurrency concerns and the complete copy of the list.  But this is ideal to the scenario we have because the system transformers should be long lived services that get registered early in startup (before any AppClassLoader is even created) and then only unregistered on shutdown after the AppClassLoader has stopped being used by the application.

> 
> There are a few copyrights that need update; Libby will whine about that.

I'll fix that, although in past experience I've noticed no warning at all from Libby about copyright dates.

> 
> Lastly, my understanding is that a beta fence should throw an exception when running in non-beta images and emit a single info message in beta images. This solution enables new behavior for beta images, only, which I think is sufficient as the beta fence I described should be added to the feature/service which uses the whiteboard.

I am not aware of such a policy where an exception must be thrown.  I thought it was sufficient to simply ignore the usage.  Besides that, I see no good way to throw an exception simply because some bundle registers a ClassFileTransformer service with some specific service properties.  I suppose we could easily log a warning/error.  But that seems unnecessary.  I am also pretty sure any beta configuration that is specified by the customer is simply ignored and no warnings/errors are produced.  Regardless, if you have the policy handy could you link me to it?

> 
> Approved since I found no required changes other than copyrights, which libby will handle.

tjwatson committed:
Updates from review

- Update copyright dates
- Update debug message to use "system" transformer instead of "global"

tjwatson commented:
#build

tjwatson commented:
#build

jhanders34 commented on dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java:
```diff
@@ -62,6 +62,7 @@
 import com.ibm.ws.ffdc.annotation.FFDCIgnore;
 import com.ibm.ws.kernel.boot.classloader.ClassLoaderHook;
 import com.ibm.ws.kernel.boot.classloader.ClassLoaderHookFactory;
+import com.ibm.ws.kernel.productinfo.ProductInfo;
```
Not used in this class from what I can see.

tjwatson replied:
Cleaned up the imports, there were several unnecessary imports.  This project meta-data doesn't appear to have the clean-up actions set to run on save.

jhanders34 commented on dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java:
```diff
@@ -111,14 +112,17 @@ private TraceComponent getClassLoadingTraceComponent(String pkg) {
-    AppClassLoader(ClassLoader parent, ClassLoaderConfiguration config, List<Container> containers, DeclaredApiAccess access, ClassRedefiner redefiner, ClassGenerator generator, GlobalClassloadingConfiguration globalConfig) {
+    AppClassLoader(ClassLoader parent, ClassLoaderConfiguration config, List<Container> containers, DeclaredApiAccess access, ClassRedefiner redefiner, ClassGenerator generator, GlobalClassloadingConfiguration globalConfig, List<ClassFileTransformer> systemTransformers) {
         super(containers, parent, redefiner, globalConfig);
+        // only use system transformers if IS_BETA (part of checkpoint feature)
```
Comment is not accurate to this class since there is no IS_BETA field in this class.

tjwatson replied:
I removed the comment.  I considered modifying the comment to state that the collection will be empty if not in beta.  But I figured such a comment will simply become stale when we forget to remove it once this stuff moves out of beta.

jhanders34 reviewed (approved):
Two minor comments.  Approving.

tjwatson committed:
Clean up from review

- Fix up unused imports
- remove comment about beta
- remove unused variable (not part of this review directly, just
additional needed clean up).

tjwatson commented:
#build

One final build to verify after a rebase.

tjwatson commented:
#libby
